### PR TITLE
New: Automatically add artists found in a root folder

### LINKED ISSUE #207: Possible to add albums with "Various Artists" types?
I like soundtracks and broadway cast albums. Example https://musicbrainz.org/release/19e87c02-28df-4a89-ba7e-8f5ef474c36f

Is it possible to add these types as they don't have a set "Artist"

### LINKED ISSUE #467: Soundtracks as a first-class object
**Why bother?**
The issue is, first and foremost, that there is no proper way to find/track/download OSTs for movies, games or anime. And I already have quite a collection. It gets even worse when I'm trying to plunge it into Plex. So, 1) find and download, 2) integrate into Emby/Plex/whatever.

**My take on automatized solution**
Since the central issue is that soundtracks are, basically, a music collections and "fall apart" when you try to bind them to Artist, I propose to categorize them differently in terms of Lidarr entities:

1. Instead of [Artist] use [Product] (such as "Cowboy Bebop", "Silent Hill 3" or "The Dark Knight")
2. [Album] - same as in case with any other music
3. [Track] - override with "[Artist] - [Track]" (would be nice to overwrite tags for library!)

I'm not yet familiar with the code, but I expect something like a inherited "Soundtrack" from "Artist". That's a thing subject to research.

Also, I briefly searched for data sources I could use for metadata.

https://vgmdb.net - looks very promising for anime and games. I tried few queries and am greatly satisfied with it
http://www.soundtrackinfo.com/ - should do the trick for movies

**Alternatives?**
Picard or Mp3Tag helps a great deal, but fall short on streamlining the process part.

### LINKED ISSUE #558: Recognize Artist from Single Tracks
Hi, 
My Music Library Structure looks like This, I have a folder for each Letter, so A, B, C, D and so on, and inside those Folders I have all music files like Artist - Song.mp3 as example while Artist A belongs into folder A and so on. My issue now is, I can't as example choose Folder A and let it "scrape" there. Would be great if there is an option to check files without folders.
Thanks

### LINKED ISSUE #752: Allow importing albums of same artist from multiple root folders
**Is your feature request related to a problem? Please describe.**
I have two root folders, one in which I store archive copies of albums, consisting mostly of single file albums (e.g. FLAC, APE, ISO) with cue files.
The other contains downloaded albums in varying quality.
If I try adding the `archive` root folder, it won't import the artists that also exist in the `dynamic` folder. So in the final artist view, those albums existing in the `archive` folder are still listed as 'missing'

**Describe the solution you'd like**
I would like to be able to import/scan albums where the artists already exists in an earlier added rootfolder.

**Additional context**
Currently even if these artist albums would be allowed to import, it wouldn't work since lidarr can't parse single file albums. see issue  #515)

### LINKED ISSUE #792: Double-Artist album (aka split) handling
Hi all,

I am new to Lidarr (swapping over from Headphones) and this is more a question than a bug report.

My library is consisting mostly out of Hardcore/Punk music and I am not pretty sure if this is somewhat special to that kind of music (never was into different stuff). There are so called split albums which contain tracks from 2 different artists. They are not handeld/called as various artists albums. (e.g.: https://musicbrainz.org/release-group/a0868611-5c89-3b1a-a29d-57278a80bb7e a split from $wingin' Utter$ and Youth Brigade).

Musicbrainz shows them like that:

Bildschirmfoto 2019-04-28 um 08 44 56

Bildschirmfoto 2019-04-28 um 08 45 14

My library was build/managed by the famous [beets](https://github.com/beetbox/beets) library manager and I actually want to keep it, because of the feature richness that tool provides.

Now to the problem I am facing:

beets is importing those split albums into a folder like `$wingin’ Utter$ _ Youth Brigade` - as a side node my folder name template is like `{Artist Name}/{Release Year} - {Album Title}` - because this is a multi artist release. The album is not putten under one of those artists nor both of them.

Bildschirmfoto 2019-04-28 um 08 46 30

and a query inside beets shows which tracks are from what artist:

Bildschirmfoto 2019-04-28 um 08 52 04

Bildschirmfoto 2019-04-28 um 08 47 21

Now when it comes to importing those albums into Lidarr I just saw that they are putten under just one artist - in the given example it's under $wingin' Utters but not under Youth Brigade:

Bildschirmfoto 2019-04-28 um 08 43 59

Youth Brigade doesn't have any sign they ever contributed to that album.

Bildschirmfoto 2019-04-28 um 08 44 31

Trying to import the mixed folder, as it was created from beets, is not an allowed while it is disabled

Bildschirmfoto 2019-04-28 um 08 46 00

So my question is: What are my options? Is there any chance to get those albums imported withou changing the file-structure? My library is quite huge and was build over couple of years with lot's of time to get everything tagged quite nice. I don't want to change much on that side... I think I do know the answer 😞 

Sorry for quite a lot of pictures - I just putten them here to visualize the problem I have.

Thanks, m.

### LINKED ISSUE #1012: Failed to recognise new path when changing root folder
**Describe the bug**

When changing the root folder using 'Mass Editor' and moving the files, the Library doesn't update its file location

**To Reproduce**
1. Go to 'Library'
2. Click on 'Mass Editor'
3. Select an Artist, select a new 'Root Folder' and select 'Yes, Move the Files'
4. Browse to the artist in your Library and see the error 'Loading track files failed' where the album list should be

**Expected behavior**

Lidarr should update its file database to use the new root

**Logs**

```
[Fatal] LidarrErrorPipeline: Request Failed. GET /api/v1/trackFile

[v0.7.1.1381] NzbDrone.Common.Disk.NotParentException: /[OLD_ROOT_FOLDER]/[ARTIST]/[ALBUM]/[TRACK] is not a child of /[NEW_ROOT_FOLDER]/[ARTIST]
  at NzbDrone.Common.Extensions.PathExtensions.GetRelativePath (System.String parentPath, System.String childPath) [0x00021] in <96b70a41daeb490db51c9994341a8d2b>:0
  at Lidarr.Api.V1.TrackFiles.TrackFileResourceMapper.ToResource (NzbDrone.Core.MediaFiles.TrackFile model, NzbDrone.Core.Music.Artist artist, NzbDrone.Core.DecisionEngine.Specifications.IUpgradableSpecification upgradableSpecification) [0x00047] in <2883fe04aa0a46f687bad43f547ee0c6>:0
  at Lidarr.Api.V1.TrackFiles.TrackFileModule+<>c__DisplayClass9_0.<GetTrackFiles>b__1 (NzbDrone.Core.MediaFiles.TrackFile f) [0x00000] in <2883fe04aa0a46f687bad43f547ee0c6>:0
  at System.Collections.Generic.List`1[T].ConvertAll[TOutput] (System.Converter`2[TInput,TOutput] converter) [0x0001a] in <254335e8c4aa42e3923a8ba0d5ce8650>:0
  at Lidarr.Api.V1.TrackFiles.TrackFileModule.GetTrackFiles () [0x00ad7] in <2883fe04aa0a46f687bad43f547ee0c6>:0
  at Lidarr.Http.REST.RestModule`1[TResource].<set_GetResourceAll>b__33_0 (System.Object options) [0x00006] in <26a7215ffa4844618987ce6727897136>:0
  at (wrapper dynamic-method) System.Object.CallSite.Target(System.Runtime.CompilerServices.Closure,System.Runtime.CompilerServices.CallSite,System.Func`2<object, object>,object)
  at Nancy.Routing.Route+<>c__DisplayClass4.<Wrap>b__3 (System.Object parameters, System.Threading.CancellationToken context) [0x00049] in <bb3027f50b35411088f45475912cc2ff>:0
```

**System info:**
 - Lidarr Version: 0.7.1.1381
 - Operating System: CoreOS Container Linux
 - mono Version: 5.20.1.34

**Additionl info:**

I am using the Docker container [linuxserver/lidarr](https://hub.docker.com/r/linuxserver/lidarr)

### LINKED ISSUE #1068: Import page broken
**Describe the bug**
Import page does not load and throws error in browser console

**To Reproduce**
1. Go to Import (under Library)
2. Click Path
3. Blank screen (Error in console)

**Expected behavior**
Something displayed, whether unmapped directories or message stating all have been imported.

**Logs**
Nothing in the application logs, this is the error in the browser console:
```
breadcrumbs.js:58
TypeError: Cannot read property 'defaults' of undefined
    at ImportArtistConnector.js:46
    at index.js:70
    at index.js:30
    at index.js:84
    at Function.mapToProps (index.js:30)
    at r (wrapMapToProps.js:41)
    at Function.r.mapToProps (wrapMapToProps.js:55)
    at r (wrapMapToProps.js:41)
    at selectorFactory.js:22
    at connectAdvanced.js:210
    at Object.useMemo (react-dom.production.min.js:159)
    at useMemo (react.production.min.js:23)
    at h (connectAdvanced.js:195)
    at $o (react-dom.production.min.js:150)
    at Di (react-dom.production.min.js:166)
    at ki (react-dom.production.min.js:166)
    at Li (react-dom.production.min.js:165)
    at Hi (react-dom.production.min.js:184)
    at qa (react-dom.production.min.js:232)
    at Ka (react-dom.production.min.js:233)
    at Ys (react-dom.production.min.js:249)
    at Ps (react-dom.production.min.js:248)
    at Rs (react-dom.production.min.js:251)
    at Dn (react-dom.production.min.js:85)
    at HTMLDocument.r (helpers.js:74)
```

**System info (please complete the following information):**
 - Lidarr Version: 0.7.1.1612
 - Operating System: docker
 - .net Framework (Windows) or mono (macOS/Linux) Version: 5.20.1.34

### PR DESCRIPTION
**Pull request by ta264**
#### Database Migration
YES

#### Description
Point lidarr at a root folder and let it rip.  It'll automatically match up the albums found and add artists with the specified options.

#### Todos
- [x] Tests
- [x] Documentation

#### Issues Fixed or Closed by this PR

* Fixes #207 
* Fixes #752 
* Fixes #1012 
* Fixes #1068 
* Fixes #792 
* Fixes #467 
* Fixes #558

### CODE REVIEW

File: src/NzbDrone.Core/Tags/TagService.cs [hunk 1]
```diff
@@ -109,7 +115,8 @@ public List<TagDetails> Details()
                         ImportListIds = importLists.Where(c => c.Tags.Contains(tag.Id)).Select(c => c.Id).ToList(),
                         NotificationIds = notifications.Where(c => c.Tags.Contains(tag.Id)).Select(c => c.Id).ToList(),
                         RestrictionIds = restrictions.Where(c => c.Tags.Contains(tag.Id)).Select(c => c.Id).ToList(),
-                        ArtistIds = artists.Where(c => c.Tags.Contains(tag.Id)).Select(c => c.Id).ToList()
+                        ArtistIds = artists.Where(c => c.Tags.Contains(tag.Id)).Select(c => c.Id).ToList(),
+                        RootFolderIds = rootFolders.Where(c => c.DefaultTags.Contains(tag.Id)).Select(c => c.Id).ToList()
```
Comment (Qstick) [inline src/NzbDrone.Core/Tags/TagService.cs]: Need to also update `QualityProfileService.cs` and `MetadataProfileService.cs`. Both of their `Delete` functions check for in use prior to allowing deletion. Delete should be prevented if a root folder is using them.

File: src/Lidarr.Api.V1/RootFolders/RootFolderModule.cs [hunk 2]
```diff
@@ -27,17 +33,29 @@ FolderWritableValidator folderWritableValidator
...
+                .SetValidator(rootFolderValidator);
+
+            SharedValidator.RuleFor(c => c.Name)
+                .NotEmpty();
+
+            SharedValidator.RuleFor(c => c.DefaultMetadataProfileId)
+                .SetValidator(metadataProfileExistsValidator);
+
+            SharedValidator.RuleFor(c => c.DefaultQualityProfileId)
+                .SetValidator(profileExistsValidator);
```
Comment (Qstick) [inline src/Lidarr.Api.V1/RootFolders/RootFolderModule.cs]: Rename to `qualityProfileExistsValidator`

**Comment by hikaricore** [comment]
I'm not sure if this is the right place to be asking about this, but I had a few questions since I missed out on testing this before and thought asking would be easier than digging through the code to try finding out myself.  XD

1. Is this feature enabled by default?
2. Can it even be disabled?
3. Is is possible to exclude directories within the root folder such as Various Artists ( https://musicbrainz.org/artist/89ad4ac3-39f7-470e-963a-56509c546377 ) which could result in a ton of unintended or incorrect additions.

My apologies if posting my questions here is uncouth.  ^_^

**Comment by ta264** [comment]
1. Yes
2. No
3. No - and it won't result in incorrect additions I don't think. For certain artists, like va, we only add matching albums and don't allow all albums to be added.

**Pull request merged**